Capturing IRC error "Login authentication failed"

Thoughts on capturing the "Login authentication failed" error and returning and error from Connect() if we jump into that error? Would make it a bit more obvious if we accidentally use the wrong oauth token

@gempir 
I have an implementation that I use locally atm but I would like to know if it fits the library before I bother to clean up the code and PR

Example usage code (note that nothing changed, we just propagate the error to `Connect`):
```go
package main

import (
	twitch "github.com/gempir/go-twitch-irc"
)

func main() {
	client := twitch.NewClient("adaljap", "penis")

	client.OnNewMessage(func(channel string, user twitch.User, message twitch.Message) {
	})

	client.Join("pajlada")

	err := client.Connect()
	if err != nil {
		panic(err)
	}
}
```

Console output:
```sh
$ go run main.go
panic: login authentication failed

goroutine 1 [running]:
main.main()
        C:/dev/go/src/github.com/pajlada/twitchtags/main.go:17 +0x25e
exit status 2
```

--- a/client.go
+++ b/client.go
@@ -22,6 +22,9 @@ const (
 var (
 	// ErrClientDisconnected returned from Connect() when a Disconnect() was called
 	ErrClientDisconnected = errors.New("client called Disconnect()")
+
+	// ErrLoginAuthenticationFailed returned from Connect() when either the wrong or a malformed oauth token is used
+	ErrLoginAuthenticationFailed = errors.New("login authentication failed")
 )
 
 // User data you receive from tmi
@@ -231,6 +234,9 @@ func (c *Client) Connect() error {
 
 		err = c.readConnection(c.connection)
 		if err != nil {
+			if err == ErrLoginAuthenticationFailed {
+				return err
+			}
 			time.Sleep(time.Millisecond * 200)
 			continue
 		}
@@ -271,7 +277,9 @@ func (c *Client) readConnection(conn net.Conn) error {
 					c.onConnect()
 				}
 			}
-			c.handleLine(msg)
+			if err = c.handleLine(msg); err != nil {
+				return err
+			}
 		}
 	}
 }
@@ -304,10 +312,15 @@ func (c *Client) send(line string) {
 	}
 }
 
-func (c *Client) handleLine(line string) {
+// Errors returned from handleLine break out of readConnections, which starts a reconnect
+// This means that we should only return fatal errors as errors here
+func (c *Client) handleLine(line string) error {
 	if strings.HasPrefix(line, "PING") {
 		c.send(strings.Replace(line, "PING", "PONG", 1))
+
+		return nil
 	}
+
 	if strings.HasPrefix(line, "@") {
 		channel, user, clientMessage := ParseMessage(line)
 
@@ -345,7 +358,10 @@ func (c *Client) handleLine(line string) {
 				c.onNewUnsetMessage(clientMessage.Raw)
 			}
 		}
+
+		return nil
 	}
+
 	if strings.HasPrefix(line, ":") {
 		if strings.Contains(line, "tmi.twitch.tv JOIN") {
 			channel, username := parseJoinPart(line)
@@ -379,7 +395,14 @@ func (c *Client) handleLine(line string) {
 				c.channelUserlist[channel][user] = true
 			}
 		}
+		if strings.Contains(line, "tmi.twitch.tv NOTICE * :Login authentication failed") || strings.Contains(line, "tmi.twitch.tv NOTICE * :Improperly formatted auth") {
+			return ErrLoginAuthenticationFailed
+		}
+
+		return nil
 	}
+
+	return nil
 }
 
 // ParseMessage parse a raw ircv3 twitch

--- a/client.go
+++ b/client.go
@@ -398,8 +398,6 @@ func (c *Client) handleLine(line string) error {
 		if strings.Contains(line, "tmi.twitch.tv NOTICE * :Login authentication failed") || strings.Contains(line, "tmi.twitch.tv NOTICE * :Improperly formatted auth") {
 			return ErrLoginAuthenticationFailed
 		}
-
-		return nil
 	}
 
 	return nil

--- a/client_test.go
+++ b/client_test.go
@@ -36,6 +36,46 @@ func newTestClient(host string) *Client {
 	return client
 }
 
+func handleTestConnection(t *testing.T, onConnect func(net.Conn), onMessage func(string), listener net.Listener) {
+	conn, err := listener.Accept()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	reader := bufio.NewReader(conn)
+	tp := textproto.NewReader(reader)
+
+	defer listener.Close()
+	defer conn.Close()
+	for {
+		message, err := tp.ReadLine()
+		if err != nil && err != io.EOF {
+			t.Fatal(err)
+		}
+		message = strings.Replace(message, "\r\n", "", 1)
+
+		if strings.HasPrefix(message, "NICK") {
+			fmt.Fprintf(conn, ":tmi.twitch.tv 001 justinfan123123 :Welcome, GLHF!\r\n")
+			onConnect(conn)
+			continue
+		}
+
+		if strings.HasPrefix(message, "PASS") {
+			pass := strings.Split(message, " ")[1]
+			if !strings.HasPrefix(pass, "oauth:") {
+				fmt.Fprintf(conn, ":tmi.twitch.tv NOTICE * :Improperly formatted auth\r\n")
+				return
+			} else if pass == "oauth:wrong" {
+				fmt.Fprintf(conn, ":tmi.twitch.tv NOTICE * :Login authentication failed\r\n")
+				return
+			}
+		}
+
+		onMessage(message)
+	}
+
+}
+
 func startServer(t *testing.T, onConnect func(net.Conn), onMessage func(string)) string {
 	host := "127.0.0.1:" + strconv.Itoa(startPort)
 	startPort++
@@ -51,31 +91,8 @@ func startServer(t *testing.T, onConnect func(net.Conn), onMessage func(string))
 	if err != nil {
 		t.Fatal(err)
 	}
-	go func() {
-		conn, err := listener.Accept()
-		if err != nil {
-			t.Fatal(err)
-		}
 
-		reader := bufio.NewReader(conn)
-		tp := textproto.NewReader(reader)
-
-		defer listener.Close()
-		defer conn.Close()
-		for {
-			message, err := tp.ReadLine()
-			if err != nil && err != io.EOF {
-				t.Fatal(err)
-			}
-			message = strings.Replace(message, "\r\n", "", 1)
-			if strings.HasPrefix(message, "NICK") {
-				fmt.Fprintf(conn, ":tmi.twitch.tv 001 justinfan123123 :Welcome, GLHF!\r\n")
-				onConnect(conn)
-			} else {
-				onMessage(message)
-			}
-		}
-	}()
+	go handleTestConnection(t, onConnect, onMessage, listener)
 
 	return host
 }
@@ -88,31 +105,8 @@ func startNoTLSServer(t *testing.T, onConnect func(net.Conn), onMessage func(str
 	if err != nil {
 		t.Fatal(err)
 	}
-	go func() {
-		conn, err := listener.Accept()
-		if err != nil {
-			t.Fatal(err)
-		}
 
-		reader := bufio.NewReader(conn)
-		tp := textproto.NewReader(reader)
-
-		defer listener.Close()
-		defer conn.Close()
-		for {
-			message, err := tp.ReadLine()
-			if err != nil && err != io.EOF {
-				t.Fatal(err)
-			}
-			message = strings.Replace(message, "\r\n", "", 1)
-			if strings.HasPrefix(message, "NICK") {
-				fmt.Fprintf(conn, ":tmi.twitch.tv 001 justinfan123123 :Welcome, GLHF!\r\n")
-				onConnect(conn)
-			} else {
-				onMessage(message)
-			}
-		}
-	}()
+	go handleTestConnection(t, onConnect, onMessage, listener)
 
 	return host
 }
@@ -167,7 +161,12 @@ func TestCanConnectAndAuthenticate(t *testing.T) {
 
 	client := NewClient("justinfan123123", oauthCode)
 	client.IrcAddress = host
-	go client.Connect()
+	go func() {
+		err := client.Connect()
+		if err != nil {
+			t.Fatal("bad error")
+		}
+	}()
 
 	select {
 	case <-wait:
@@ -774,6 +773,28 @@ func TestCanNotDialInvalidAddress(t *testing.T) {
 	err := client.Connect()
 	if !strings.Contains(err.Error(), "invalid port") {
 		t.Fatal("invalid Connect() error")
+	}
+}
+
+func TestCanNotUseImproperlyFormattedOauth(t *testing.T) {
+	host := startServer(t, nothingOnConnect, nothingOnMessage)
+	client := NewClient("justinfan123123", "imrpproperlyformattedoauth")
+	client.IrcAddress = host
+
+	err := client.Connect()
+	if err != ErrLoginAuthenticationFailed {
+		t.Fatal("wrong Connect() error")
+	}
+}
+
+func TestCanNotUseWrongOauth(t *testing.T) {
+	host := startServer(t, nothingOnConnect, nothingOnMessage)
+	client := NewClient("justinfan123123", "oauth:wrong")
+	client.IrcAddress = host
+
+	err := client.Connect()
+	if err != ErrLoginAuthenticationFailed {
+		t.Fatal("wrong Connect() error")
 	}
 }
 

